[Unified Text Replacement] Text replacements do not have their associated document markers properly updated

rr-codes opened the pull request:
#### 3a919dcf057367c9e1ccc0b9be32f581989d7a4e
<pre>
[Unified Text Replacement] Text replacements do not have their associated document markers properly updated
<a href="https://bugs.webkit.org/show_bug.cgi?id=268886">https://bugs.webkit.org/show_bug.cgi?id=268886</a>
<a href="https://rdar.apple.com/120563596">rdar://120563596</a>

Reviewed by Aditya Keerthi.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::description const):
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/Shared/WebTextReplacementData.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::WebPageProxy::didEndTextReplacementSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::WebPage::didEndTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/274292@main">https://commits.webkit.org/274292@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee17926c0af792e6603e1ba978c164b2bdb93d3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38640 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17569 "Failed to checkout and rebase branch from PR 23977") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/40965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/41170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/40945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20374 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/14916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/41170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39213 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/14809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/40965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/41170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/40965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/42445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/35103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/40965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/42445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/14916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/42445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/40965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5028 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -40,6 +40,8 @@
 #include <WebCore/PlatformTextAlternatives.h>
 #endif
 
+#include "WebTextReplacementData.h"
```
I think this can be forward declared. We really want to minimize includes in this file.

rr-codes replied:
yup, I originally thought it was needed because of the nested enum, but now that I think about it the enum isn’t actually nested

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -53,6 +54,28 @@ static void replaceTextInRange(WebCore::LocalFrame& frame, const SimpleRange& ra
     frame.editor().replaceSelectionWithText(replacementText, WebCore::Editor::SelectReplacement::Yes, WebCore::Editor::SmartReplace::No, WebCore::EditAction::InsertReplacement);
 }
 
+static std::optional<std::tuple<Node&, DocumentMarker&>> findReplacementMarkerByUUID(WebCore::Document* document, const WTF::UUID& replacementUUID)
```
This should take a reference to `Document`.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -53,6 +54,28 @@ static void replaceTextInRange(WebCore::LocalFrame& frame, const SimpleRange& ra
+static std::optional<std::tuple<Node&, DocumentMarker&>> findReplacementMarkerByUUID(WebCore::Document* document, const WTF::UUID& replacementUUID)
+{
+    Node* targetNode { nullptr };
+    DocumentMarker* targetMarker { nullptr };
```
These should be smart pointers.

rr-codes replied:
what kind should they be?

cdumez replied:
Node is RefCounted so RefPtr is suitable for stack variables. DocumentMarker is not refcounted though so best we can do is WeakPtr at the moment.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+    if (state != WebTextReplacementData::State::Committed && state != WebTextReplacementData::State::Reverted)
+        return;
+
+    auto nodeAndMarker = findReplacementMarkerByUUID(document.get(), replacement.uuid);
```
You've already null checked document, so you can pass a reference here.

rr-codes replied:
as in `*document.get()`? Or pass in a `Ref` as `document`?

pxlcoder replied:
`*document`

pxlcoder replied:
Passing a `Ref` would introduce unwanted ref-churn.

pxlcoder replied:
Just to clarify, `*document` is a shorter way to write `*document.get()`.

rr-codes replied:
ohhhhh I see!

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+        }
+    }();
+
+    replaceTextInRange(frame.get(), rangeToReplace, newText);
```
Just to confirm, this remove the old marker right?

rr-codes replied:
it actually probably does, but I already explicitly remove the markers anyways on line 213

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+        return;
+    }
+
+    // FIXME: Associate the markers with a specific session, and only modify the markers which belong
+    // to this session.
```
This is an optimization, and not a correctness `FIXME` right?

In any case, a filing a bug would be good.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -148,6 +148,8 @@
 #include <WebCore/MediaSessionIdentifier.h>
 #endif
 
+#include "WebTextReplacementData.h"
```
Can we avoid this include?

rr-codes replied:
probably!

cdumez commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+        return;
+    }
+
+    auto* corePage = m_webPage->corePage();
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
+        return;
+    }
+
+    auto* corePage = m_webPage->corePage();
```
RefPtr

aprotyas commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -53,6 +54,28 @@ static void replaceTextInRange(WebCore::LocalFrame& frame, const SimpleRange& ra
+        return true;
+    });
+
+    if (targetNode && targetMarker)
+        return { { *targetNode, *targetMarker } };
```
Oh this ended up working out! 😄

aprotyas reviewed (commented):
Looks good, echoing most of the existing comments!

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -53,6 +54,28 @@ static void replaceTextInRange(WebCore::LocalFrame& frame, const SimpleRange& ra
+static std::optional<std::tuple<Node&, DocumentMarker&>> findReplacementMarkerByUUID(const WebCore::Document& document, const WTF::UUID& replacementUUID)
+{
+    RefPtr<Node> targetNode { nullptr };
+    WeakPtr<DocumentMarker> targetMarker { nullptr };
```
`{ nullptr }` should not be necessary.

rr-codes committed:
[Unified Text Replacement] Text replacements do not have their associated document markers properly updated
https://bugs.webkit.org/show_bug.cgi?id=268886
rdar://120563596

Reviewed by Aditya Keerthi.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::description const):
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in:
* Source/WebKit/Shared/WebTextReplacementData.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::WebPageProxy::didEndTextReplacementSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::WebPage::didEndTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::findReplacementMarkerByUUID):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/274292@main
